### PR TITLE
Reconnect automatically and re-wire the sockets on connection lost

### PR DESCRIPTION
**Pull request by 0cv**
Solve the point 1/ of https://github.com/adampointer/go-deribit/issues/21

### CODE REVIEW

File: v3/auth.go [hunk 1]
```diff
@@ -10,9 +10,19 @@ import (
...
 const renewBefore int64 = 600
 
-func (e *Exchange) Authenticate(key, secret string) error {
+// Keep clientID and clientSecret for re-connecting automatically
+var (
+	clientID     string
+	clientSecret string
+)
+
+func (e *Exchange) Authenticate(keys ...string) error {
```
Comment (adampointer) [inline v3/auth.go:20]: Can you add a docblock to explain what arguments need to be passed here please? Just so users do not need to read the code to understand how to use the method.

File: v3/exchange.go [hunk 2]
```diff
@@ -87,7 +88,33 @@ func (e *Exchange) heartbeat() {
 			select {
 			case <-ticker.C:
 				if _, err := e.Client().GetPublicTest(&operations.GetPublicTestParams{}); err != nil {
-					e.stop <- true
+
```
Comment (adampointer) [inline v3/exchange.go]: Split this out into a new method `Reconnect`

File: v3/exchange.go [hunk 3]
```diff
@@ -87,7 +88,33 @@ func (e *Exchange) heartbeat() {
...
+					// We lost the connection so we attempt to re-connect
+					c, _, err := websocket.DefaultDialer.Dial(e.url, nil)
+					if err != nil {
+						log.Printf("Error in the default dialer %v", err)
```
Comment (adampointer) [inline v3/exchange.go]: Add a method on exchange to allow users to redirect the output of this if they want.

e.g.

```
func (e *Exchange) SetLogOutput(w io.Writer) {
    log.SetOutput(w)
} 
```

This way users can redirect the log output to an existing logrus (or other logging package) instance.

**Review by adampointer — COMMENTED** [review]
Nice, just a few tweaks as explained in your issue.

**Comment by adampointer** [comment]
LGTM

**Comment by 0cv** [comment]
Ok I'm little unsure with the `read()` method here. I had to keep the check in the hearbeat as otherwise the `read()` method was exiting with the break Loop here: https://github.com/adampointer/go-deribit/pull/22/files#diff-000c2bb3a828c114a7a4944ad4e335a9R106. Removing the Break loop, and replacing it with e.g. `continue` was making matter worse as it came into an infinite queue of reconnect... So, it looks like it works, but please have a look

**Comment by 0cv** [comment]
Oh good you're fast, then fine :-)

**Comment by adampointer** [comment]
That read method is a bit of a mess tbh.

**Pull request merged**